### PR TITLE
Use latest release from sqlcmd

### PR DESCRIPTION
**Pull request by zijchen**
Use latest sqlcmd version from https://github.com/microsoft/go-sqlcmd/releases/latest

### CODE REVIEW

**Comment by zijchen** [comment]
Should we introduce a new `sqlcmd-version` input?

**Comment by dzsquared** [comment]
> Should we introduce a new `sqlcmd-version` input?
On a similar topic - I think we want to have an implementation in the future that creates a code path where sqlcmd isn't downloaded unless its missing.
At the very least, this will save time in the runner. This also sets environments up to have reproducible sets of software.

**Review by dzsquared — APPROVED** [review]

File: src/Setup.ts [hunk 1]
```diff
@@ -2,35 +2,64 @@
...
         // Add sqlcmd to PATH
         core.addPath(sqlcmdPath);
     }
 
+    /**
+     * Gets the version number from the latest release based on the redirected URL.
+     * @returns The version number or undefined if the version could not be parsed or HTTP request failed.
+     */
+    public static async extractVersionFromLatestRelease(releaseUrl: string): Promise<string | undefined> {
+        const http = new HttpClient('Azure/sql-action', undefined, { allowRedirects: false });
```
Comment (chlafreniere) [inline src/Setup.ts:46]: Do we have access to user token from github action? Since this is unauthenticated, it will be much more rate-limited.

If we do, may be easier just to use octokit library instead of httpclient

Comment (zijchen) [reply src/Setup.ts:46]: Unfortunately it doesn't appear there's an easy way to get the token, we'd have to introduce a new input so users can pass in `GITHUB_TOKEN` to.

Comment (dzsquared) [reply src/Setup.ts:46]: I'm not sure what the rate limiting will be here - since we're not using a REST API necessarily, we're checking the redirect.

**Review by Imran-imtiaz48 — COMMENTED** [review]
### Review of TypeScript Test Code for `Setup.ts`

#### Summary:
The code snippet tests the `setupSqlcmd` function in TypeScript, ensuring correct setup of SQL command-line tools (`sqlcmd`). It utilizes Jest for mocking and spying on methods from `HttpClient` and `tc` (Tool Cache).

#### Feedback and Improvements:

1. **Test Function Naming:**
   - Ensure descriptive names for test functions. For example, `it('sets up sqlcmd correctly')` clearly describes the purpose of the test.

2. **Spy Usage:**
   - Good use of `jest.spyOn` to mock `HttpClient.prototype.head` and various methods from `tc` for controlled testing environments.

3. **Refactoring Suggestions:**
   - Consider refactoring to improve test coverage or to handle additional edge cases specific to `setupSqlcmd`.
   - Use `beforeEach` or `beforeAll` hooks to set up common test conditions, reducing redundancy in test setup.

4. **Conditional Testing:**
   - Ensure that conditional testing (`if (process.platform === 'win32')`) is robust and handles all expected scenarios.

#### Example Improvements:

```typescript
describe('Setup.ts tests', () => {
    beforeEach(() => {
        jest.clearAllMocks(); // Reset all spies before each test
    });

    it('sets up sqlcmd correctly', async () => {
        const extractVersionSpy = jest.spyOn(Setup, 'extractVersionFromLatestRelease').mockResolvedValue('1.1.1');
        const cacheLookupSpy = jest.spyOn(tc, 'find').mockReturnValue('');
        const downloadToolSpy = jest.spyOn(tc, 'downloadTool').mockResolvedValue('');
        const extractTarSpy = jest.spyOn(tc, 'extractTar').mockResolvedValue('');

        await Setup.setupSqlcmd();

        expect(extractVersionSpy).toHaveBeenCalled();
        expect(cacheLookupSpy).toHaveBeenCalled();
        expect(downloadToolSpy).toHaveBeenCalled();
        
        if (process.platform === 'win32') {
            // Add platform-specific expectations or tests
        }
    });
});
```

#### Conclusion:
The provided test setup effectively verifies the `setupSqlcmd` function's behavior, utilizing Jest spies to mock dependencies and assert expected function calls. Consider the outlined improvements to enhance code readability, maintainability, and test coverage.
